### PR TITLE
Refactor libcurl metadata fetcher to accept RequestMessage as argument

### PR DESCRIPTION
**Pull request by suniltheta**
Signed-off-by: Sunil Narasimhamurthy <sunnrs@amazon.com>

Commit Message: Refactor libcurl metadata fetcher to accept RequestMessage as argument
Additional Description:

1. Rename `metadataFetcher` to `fetchMetadata` function. This will be marked deprecated after introducing classes by the name `MedataFetcher` & `MetadataReceiver`. 

1. Extracted out functionalities `fetchCredentialFromInstanceRole` & `extractCredentials` to reuse these functions in async model.

Risk Level: Low
Testing: Unit testing
Docs Changes: NA
Release Notes: NA
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/common/aws/credentials_provider_impl.cc [hunk 1]
```diff
@@ -94,15 +106,32 @@ void InstanceProfileCredentialsProvider::refresh() {
...
-  if (!credential_document) {
-    ENVOY_LOG(error, "Could not load AWS credentials document from the instance metadata");
-    return;
+  Http::RequestMessageImpl message;
+  message.headers().setScheme(Http::Headers::get().SchemeValues.Http);
+  message.headers().setMethod(Http::Headers::get().MethodValues.Get);
+  message.headers().setHost(EC2_METADATA_HOST);
+  message.headers().setPath(credential_path);
+
+  if (!async) {
```
Comment (abaptiste) [inline source/extensions/common/aws/credentials_provider_impl.cc]: Could we ignore async if it's not used so that we don't introduce a bug?

File: source/extensions/common/aws/utility.cc [hunk 2]
```diff
@@ -223,19 +223,18 @@ static size_t curlCallback(char* ptr, size_t, size_t nmemb, void* data) {
...
+absl::optional<std::string> Utility::fetchMetadata(Http::RequestMessage& message) {
   static const size_t MAX_RETRIES = 4;
   static const std::chrono::milliseconds RETRY_DELAY{1000};
   static const std::chrono::seconds TIMEOUT{5};
 
   CURL* const curl = curl_easy_init();
-  if (!curl) {
-    return absl::nullopt;
-  };
+  RELEASE_ASSERT(curl != nullptr, "");
```
Comment (abaptiste) [inline source/extensions/common/aws/utility.cc]: I think we should go back to the return.  Granted we cannot proceed if curl isn't initialized, though I feel taking envoy down entirely is a bit harsh.

File: source/extensions/common/aws/utility.cc [hunk 3]
```diff
@@ -223,19 +223,18 @@ static size_t curlCallback(char* ptr, size_t, size_t nmemb, void* data) {
...
+  const auto host = message.headers().getHostValue();
+  const auto path = message.headers().getPathValue();
+  const auto scheme = message.headers().getSchemeValue();
 
-  const std::string url = fmt::format("http://{}/{}", host, path);
+  const std::string url = fmt::format("{}://{}{}", scheme, host, path);
```
Comment (abaptiste) [inline source/extensions/common/aws/utility.cc]: Would "https" or "ws" be a thing here?  I get your intent, just wondering if it's overkill :)

File: test/common/http/header_map_impl_test.cc [hunk 4]
```diff
@@ -424,9 +424,8 @@ TEST_P(HeaderMapImplTest, AllInlineHeaders) {
     INLINE_REQ_RESP_STRING_HEADERS(TEST_INLINE_STRING_HEADER_FUNCS)
   }
   {
-    // No request trailer O(1) headers.
-  }
-  {
+      // No request trailer O(1) headers.
+  } {
```
Comment (abaptiste) [inline test/common/http/header_map_impl_test.cc]: This seems unrelated.  Can we revert this change to eliminate unnecessary diffs?  Or git checkout  the main version and call it a day.

File: test/extensions/common/aws/credentials_provider_impl_test.cc [hunk 5]
```diff
@@ -62,35 +62,72 @@ TEST_F(EvironmentCredentialsProviderTest, NoSessionToken) {
...
 public:
   InstanceProfileCredentialsProviderTest()
       : api_(Api::createApiForTest(time_system_)),
-        provider_(*api_,
-                  [this](const std::string& host, const std::string& path,
-                         const std::string& auth_token) -> absl::optional<std::string> {
-                    return this->fetcher_.fetch(host, path, auth_token);
-                  }) {}
+        provider_(*api_, [this](Http::RequestMessage& message) -> absl::optional<std::string> {
+          return this->fetcher_.fetch(message);
```
Comment (abaptiste) [inline test/extensions/common/aws/credentials_provider_impl_test.cc]: Could we rename `fetcher_` to be a bit more descriptive?  Assuming this is the metadata fetcher.

**Review by abaptiste — CHANGES_REQUESTED** [review]
Minor comments.  Overall LGTM.

File: source/extensions/common/aws/utility.cc (hunk 3 above)
Comment (suniltheta) [reply source/extensions/common/aws/utility.cc]: As of now the scheme will just be `http` until we introduce the WebIdentity Credentials Provider, where it will be `https`.
As you suggest I will remove setting scheme part for now. We can reintroduce setting scheme along with WebIdentity Credentials provider. Which will be after we do the http async client part.

File: source/extensions/common/aws/utility.cc (hunk 2 above)
Comment (suniltheta) [reply source/extensions/common/aws/utility.cc]: ack

File: source/extensions/common/aws/credentials_provider_impl.cc (hunk 1 above)
Comment (suniltheta) [reply source/extensions/common/aws/credentials_provider_impl.cc]: Got it. I will remove the if-else.

File: test/common/http/header_map_impl_test.cc (hunk 4 above)
Comment (suniltheta) [reply test/common/http/header_map_impl_test.cc]: courtesy of clang formatter

File: test/extensions/common/aws/credentials_provider_impl_test.cc (hunk 5 above)
Comment (suniltheta) [reply test/extensions/common/aws/credentials_provider_impl_test.cc]: ack

File: source/extensions/common/aws/credentials_provider.h [hunk 6]
```diff
@@ -21,9 +21,10 @@ namespace Aws {
  */
 class Credentials {
 public:
-  Credentials(absl::string_view access_key_id = absl::string_view(),
-              absl::string_view secret_access_key = absl::string_view(),
-              absl::string_view session_token = absl::string_view()) {
+  explicit Credentials(absl::string_view access_key_id = absl::string_view(),
+                       absl::string_view secret_access_key = absl::string_view(),
+                       absl::string_view session_token = absl::string_view()) {
+    // TODO(suniltheta): Move credential expiration date in here
```
Comment (abaptiste) [inline source/extensions/common/aws/credentials_provider.h:27]: How would this expiration surface?  I'm not quite sure I get this comment.

Comment (suniltheta) [inline source/extensions/common/aws/credentials_provider.h:27]: I just copied this TODO from our local patch. If we have been not depending on this TODO for a while I guess I can remove it in subsequent PRs
